Internal error when taking a reference to `this` in `init()` of class/record

### Summary of Problem
In the following code, I am trying to take a reference to `this` in` init()` of a class (or record) to "rename" `this` to some other name (e.g., self), but the compilation fails with internal error. On the other hand, if I take a reference to `this` in `postinit()`, the code compiles successfully.

(This issue is not blocking for me and just reported here because of internal error.)

### Steps to Reproduce

**Source Code:**
This code results in internal error (for both class and record).

```chapel
class Cls    // or, record Rec
{
    var n: int;
    proc init() {
        const ref self = this;
        self.n = 100;
    }
}
var a = new owned Cls();
writeln( a.n );

$ chpl test.chpl
test.chpl:4: internal error: UTI-MIS-0602 chpl version 1.19.0
Note: This source location is a guess.
```

but if I change `init()` to `postinit()`, the code compiles and prints 100.

For comparison, I've tried a similar C++ code below, which seems to be working

```c++
#include <iostream>
using namespace std;

struct Cls
{
    int n;
    Cls() {
        auto& self = *this;
        self.n = 100;
    }
};
int main() {
    cout << Cls().n << endl;  // prints 100
}
```
while a similar Swift code below results in an error:

```swift
class Cls
{
    var n: Int
    init() {
        let my = self
        my.n = 100
    }
}
var a = Cls()
print( a.n )
```
If I further insert `n = 10` at the beginning of `init()` (so as to complete the initialization of properties), the code compiles and prints 100.

**Compile command:**
`chpl test.chpl`

### Configuration Information

- Output of `chpl --version`:
chpl version 1.19.0
Copyright (c) 2004-2019, Cray Inc.  (See LICENSE file for more details)

- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none *
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_AUX_FILESYS: none

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0

@bradcray @benharsh
Thank you very much for the comments, and I agree that taking a reference to `this` in `init()` should be a compile-time error if things can become unsafe/subtle/complicated. Indeed, in the code snippet above, Swift also seems to consider `let my = self` as illegal because the initialization of `self` is not completed:

```swift
test.swift:3:9: note: 'self.n' not initialized
    var n: Int
        ^
test.swift:6:14: error: use of 'self' in property access 'n' before all stored properties are initialized
        my.n = 100
             ^
test.swift:5:13: error: variable 'self.n' used before being initialized
        let my = self
            ^
```
The reason why I tried the above code is that I sometimes use a different name for the "this" variable in Python or Fortran (e.g., "me" or "my") to make the code somewhat shorter / more readable. (In those two languages, the first argument is passed explicitly to a method, so I can use any variable name.)

Now, I'm wondering if once the `init()` is completed,  is it possible (legal) to take such a reference to `this` in user-defined methods (other than `init()`)? For example,
```chapel
proc Mytype.myfunc() {
    const ref my = this;
    // math expressions using my.foo, my.baa, etc
} 
```
But if this can cause possible unsafe behaviors or more complications, I will definitely stick to "this" (no problem at all here).

--- a/compiler/passes/InitNormalize.cpp
+++ b/compiler/passes/InitNormalize.cpp
@@ -943,26 +943,34 @@ void ProcessThisUses::visitSymExpr(SymExpr* node) {
   DefExpr* field = NULL;
 
   if (node->symbol()->hasFlag(FLAG_ARG_THIS)) {
-    CallExpr* call = NULL;
-    Expr* cur = node;
-    while (call == NULL && cur->parentExpr != NULL) {
-      if (CallExpr* parent = toCallExpr(cur->parentExpr)) {
-        call = parent;
+    if (DefExpr* parentDef = toDefExpr(node->parentExpr)) {
+      if (parentDef->sym->hasFlag(FLAG_REF_VAR)) {
+        USR_FATAL_CONT(node, "cannot take a reference to \"this\" before this.complete()");
       } else {
-        cur = cur->parentExpr;
+        USR_FATAL_CONT(node, "cannot initialize a variable from \"this\" before this.complete()");
       }
-    }
-
-    if (call->isPrimitive() == false) {
-      if (state->isPhase0()) {
-        USR_FATAL_CONT(node, "cannot pass \"this\" to a function before calling super.init() or this.init()");
-      } else if (state->type()->isRecord()) {
-        USR_FATAL_CONT(node, "cannot pass a record to a function before this.complete()");
+    } else {
+      CallExpr* call = NULL;
+      Expr* cur = node;
+      while (call == NULL && cur->parentExpr != NULL) {
+        if (CallExpr* parent = toCallExpr(cur->parentExpr)) {
+          call = parent;
+        } else {
+          cur = cur->parentExpr;
+        }
       }
-    }
 
-    if (isClass(state->type())) {
-      node->setSymbol(state->getThisAsParent());
+      if (call && call->isPrimitive() == false) {
+        if (state->isPhase0()) {
+          USR_FATAL_CONT(node, "cannot pass \"this\" to a function before calling super.init() or this.init()");
+        } else if (state->type()->isRecord()) {
+          USR_FATAL_CONT(node, "cannot pass a record to a function before this.complete()");
+        }
+      }
+
+      if (isClass(state->type())) {
+        node->setSymbol(state->getThisAsParent());
+      }
     }
   } else if (DefExpr* local = state->type()->toLocalField(node)) {
     field = local;

--- a/compiler/passes/normalize.cpp
+++ b/compiler/passes/normalize.cpp
@@ -2226,6 +2226,12 @@ static void normRefVar(DefExpr* defExpr) {
     if (ArgSymbol* arg = toArgSymbol(symbol)) {
       if (arg->intent == INTENT_BLANK && arg->type == dtUnknown) {
         error = false;
+      } else if (arg->hasFlag(FLAG_ARG_THIS) &&
+                 arg->intent == INTENT_BLANK &&
+                 (arg->getFunction()->isInitializer() || arg->getFunction()->isCopyInit())) {
+        // InitNormalize.cpp will handle the case of initializing a ref-var
+        // from 'this' inside an initializer.
+        error = false;
       }
     }
 

--- a/test/classes/initializers/phase1/initFromThis.chpl
+++ b/test/classes/initializers/phase1/initFromThis.chpl
@@ -1,0 +1,42 @@
+// Based on issue #12695
+
+record R {
+  var x : int;
+
+  proc init() {
+    ref A = this;
+    A.x = 10;
+    const ref B = this;
+    writeln(B);
+    var foo = this;
+    writeln(foo);
+  }
+
+  proc init=(other:R) {
+    ref A = this;
+    A.x = other.x;
+    const ref B = this;
+    writeln(B);
+    var foo = this;
+    writeln(foo);
+  }
+}
+
+class C {
+  var x : int;
+
+  proc init() {
+    ref A = this;
+    A.x = 10;
+    const ref B = this;
+    writeln(B);
+    var foo = this;
+    writeln(foo);
+  }
+}
+
+proc main() {
+  var X : R;
+  var Y = X;
+  var Z = new C();
+}

--- a/test/classes/initializers/phase1/initFromThis.good
+++ b/test/classes/initializers/phase1/initFromThis.good
@@ -1,0 +1,12 @@
+initFromThis.chpl:6: In initializer:
+initFromThis.chpl:7: error: cannot take a reference to "this" before this.complete()
+initFromThis.chpl:9: error: cannot take a reference to "this" before this.complete()
+initFromThis.chpl:11: error: cannot initialize a variable from "this" before this.complete()
+initFromThis.chpl:15: In function 'init=':
+initFromThis.chpl:16: error: cannot take a reference to "this" before this.complete()
+initFromThis.chpl:18: error: cannot take a reference to "this" before this.complete()
+initFromThis.chpl:20: error: cannot initialize a variable from "this" before this.complete()
+initFromThis.chpl:28: In initializer:
+initFromThis.chpl:29: error: cannot take a reference to "this" before this.complete()
+initFromThis.chpl:31: error: cannot take a reference to "this" before this.complete()
+initFromThis.chpl:33: error: cannot initialize a variable from "this" before this.complete()

